Implement bidirectional_dynamic_rnn (#1779)

jihunchoi opened the pull request:
In #1779, it seems that the problem of implementing bidirectional version of `dynamic_rnn` is that `tf.reverse_sequence` cannot reverse a sequence whose shape is unknown.
However, now it can reverse an unknown-shaped sequence according to #1816, so I implemented the dynamic version of `dynamic_rnn` and its tests.

jihunchoi committed:
Implement bidirectional_dynamic_rnn (#1779)

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented:
ping for @ebrevdo

ebrevdo commented:
Will look tomorrow
On Jun 7, 2016 5:55 PM, "Vijay Vasudevan" <notifications@github.com> wrote:

> ping for @ebrevdo <https://github.com/ebrevdo>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2581#issuecomment-224457640>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABtimwGtHxN1ISsuG_2iujnK03rBfYErks5qJhMGgaJpZM4Ip5gv>
> .
>

vrv commented:
ping :)

ebrevdo commented:
Jenkins, test this please.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,109 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+          `[batch_size, max_time, cell.output_size]`.
+        If time_major == True, this will be a `Tensor` shaped:
+          `[max_time, batch_size, cell.output_size]`.
+      output_state_fw is the final state of the forward rnn.
```
document shapes of output_state_fw and _bw?

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,109 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+  output_bw = array_ops.reverse_sequence(
+      tmp, sequence_length, time_dim, batch_dim)
+  # Concat each of the forward/backward outputs
+  outputs = array_ops.concat(2, [output_fw, output_bw])
```
let's just return a tuple here.  in the documentation above, give an example showing how users can concatenate along depth with tf.concat(2, ...)

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,109 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+  # Concat each of the forward/backward outputs
+  outputs = array_ops.concat(2, [output_fw, output_bw])
+
+  return (outputs, output_state_fw, output_state_bw)
```
perhaps the output state should be a tuple as well?  first one containing fw, second one containing bw?

jihunchoi committed:
Modify return values of bidirectional_dynamic_rnn

Instead of the concatenated `Tensor` of the forward and backward
outputs, it now returns them separately as a tuple.
Additionally, the forward and backward states are now returned
as a tuple.

jihunchoi commented:
@ebrevdo Modified! It now returns `(outputs, states)`, where `outputs` is a tuple of the forward and the backward outputs, and `states` is a tuple of the forward and the backward final state.
I also modified the documentation a bit.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,113 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+    batch_dim = 1
+  with vs.variable_scope(name + "_BW") as bw_scope:
+    inputs_reverse = array_ops.reverse_sequence(
+        inputs, sequence_length, time_dim, batch_dim)
```
use keyword arguments everywhere (i.e., your call to dynamic_rnn, reverse_sequence, etc); this is both easier to read & future-proof.

jihunchoi replied:
In what cases should the keyword arguments be used?
For all parameters including those without default values, or for only parameters having default values?

`dynamic_rnn(cell_fw, inputs, sequence_length=sequence_length, initial_state=initial_state_fw...)`
`reverse_sequence(inputs, sequence_length, time_dim, batch_dim=batch_dim)`

or
`dynamic_rnn(cell=cell_fw, inputs=inputs, sequence_length=sequence_length, initial_state=initial_state_fw...)`
`reverse_sequence(inputs=inputs, seq_lengths=sequence_length, seq_dim=time_dim, batch_dim=batch_dim)`

ebrevdo replied:
good rule of thumb: if there is more than 1 parameter, pass *all* parameters you provide via keyword arguments.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,113 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+    scope: VariableScope for the created subgraph; defaults to "BiRNN"
+
+  Returns:
+    A tuple (outputs, output_states) where:
```
(outputs, final_states)

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,113 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+        in the `bidirectional_rnn`. If the concatenated one is preferred,
+        the forward and backward outputs can be concatenated as
+        `tf.concat(2, outputs)`.
+      output_states: A tuple of final states of the forward and the backward
```
A tuple (final_fw_states, final_bw_states) containing the final states...

jihunchoi replied:
I named it `output_states` to be compatible with the document of `bidirectional_rnn`. Will it be better to name it `final_states` and `(final_state_fw, final_state_bw)`?

ebrevdo replied:
Good point re: bidirectional_rnn.  Let's just be consistent.  I'm fine with output_states.

ebrevdo commented:
@jihunchoi sorry for the late response.  some more comments.

jihunchoi committed:
Enhance readability of bidirectional_dynamic_rnn

jihunchoi commented:
@ebrevdo Now it uses keyword arguments, and I refined some of the docstring.

ebrevdo commented:
Jenkins, test this please.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,120 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+        the backward rnn output `Tensor`.
+        If time_major == False (default),
+          each element will be a `Tensor` shaped:
+          `[batch_size, max_time, cell.output_size]`.
```
cell_fw.output_size, no?

jihunchoi replied:
Modified to separately saying the shapes of output_fw and output_bw.

ebrevdo commented on tensorflow/python/ops/rnn.py:
```diff
@@ -463,6 +463,120 @@ def bidirectional_rnn(cell_fw, cell_bw, inputs,
+          each element will be a `Tensor` shaped:
+          `[batch_size, max_time, cell.output_size]`.
+        If time_major == True, each element will be a `Tensor` shaped:
+          `[max_time, batch_size, cell.output_size]`.
```
ditto

jihunchoi committed:
Modified an ambiguous docstring in bidirectional_dynamic_rnn

jihunchoi committed:
Fix typo in bidirectional_dynamic_rnn

ebrevdo commented:
Jenkins, test this please.

rmlarsen commented:
@ebrevdo is this ready to merge?

ebrevdo commented:
Yes, I think so.  Thanks for your patience @jihunchoi.
